net: if: Rework interface locking

rlubos opened the pull request:
Instead of having a single global mutex, covering, in most cases
unrelated, access to all network interfaces, introduce a separate mutex
for each network interface. Details in the commit messages.

Fixes #50875

rlubos committed:
net: if: Rework interface locking

Instead of having a single global mutex, covering, in most cases
unrelated, access to all network interfaces, introduce a separate mutex
for each network interface.

Wherever interface function operates on an individual interface context
data, use local mutex protection instead. The global mutex remains only
to protect access to shared resources (IPv4/6 context pool, timer lists
etc.).

Finally, rework the timer expiry handlers, to prevent interface
functions from being called while the global lock is held (the global
mutex has to be held when iterating over the timer list, as lists are
not thread safe). The reason for this is to avoid possible deadlock
scenario. The global mutex is sometimes being locked, while inside
interface function (protected by a local mutex), so if it was allowed
the other way as well, a deadlock could occur.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: if: Improve LL address access thread safety

As `net_if_get_link_addr()` returns a pointer to the link layer
structure, some extra protection is needed to prevent its contents from
being changed while accessed. Using the mutex lock associated with an
interface should do the trick.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar reviewed (dismissed):
Nice work :+1:

carlescufi commented:
@fgrandel would you mind taking a look at this PR, since you opened the corresponding issue? thanks!

fgrandel commented:
> @fgrandel would you mind taking a look at this PR, since you opened the corresponding issue? thanks!

@carlescufi Sure - will do.

fgrandel reviewed (approved):
My questions should not block this PR. In any case this is an obvious improvement over the status quo.

This PR only partially fixes #50875, though. If the RX pool is exhausted and one thread needs TX on the same iface to unblock RX - or the other way round, a deadlock may still occur. See the detailed analysis given in https://github.com/zephyrproject-rtos/zephyr/pull/50905#issuecomment-1287864547. For this to be solved independent per-pool locks would have to be introduced.

rlubos commented on include/zephyr/net/net_if.h:
```diff
@@ -526,8 +526,24 @@ struct net_if {
 	int tx_pending;
 #endif
+
+	struct k_mutex lock;
```
> What is the performance overhead of a mutex compared to a spinlock in these cases?

I don't know to be honest, for sure this is something that could be measured to see if there's a room for a possible enhancement.

> Are we sure that none of the protected sections may ever be called from an ISR context?

That's a good point. It actually seems, that changes in `net_pkt_set_iface()` have violated this requirement. There is a short code path, between calling `net_recv_data()` and queueing the packet, when `net_pkt_set_iface()` could be called from the IRQ context (this depends on the driver design). I need to rework this func to get rid of the locking.

> So maybe a combination of two mechanisms might be adequate?

To be honest, any combnation of two or more locking mechanisms sounds like a little nightmare to me. Combining the global mutex (protecting shared resources) with the local one in this PR gave me a bit of a headache already.

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -2034,37 +2081,30 @@ struct net_if_mcast_addr *net_if_ipv6_maddr_lookup(const struct in6_addr *maddr,
 {
 	NET_ASSERT(addr);
 
-	k_mutex_lock(&lock, K_FOREVER);
```
After looking the code, I am just wondering do we really need to have a lock in maddr_leave/join functions at all.

rlubos committed:
net: iface: Improve thread safety of net_if_ipv*_maddr_* functions

Provide an extra struct net_if * iface parameter to
net_if_ipv*_maddr_join/leave functions, so that the corresponding
interface context, the mcast address belong to, can be locked for the
operation.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Updated `net_if_ipv*_maddr_join()/leave()` APIs and removed lock from `net_pkt_set_iface()`.

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -626,7 +626,7 @@ static inline bool net_if_flag_is_set(struct net_if *iface,
  *
  * @param iface Pointer to network interface
- * @param oper_state Operational state to set
+ * @param oper_state Operational state to setF
```
Is there a typo here?

jukkar reviewed (approved):
LGTM

fgrandel commented:
> The thing I don't follow, is that net_if code makes no pkt/buffer allocation so it's not clear to me where could the net_if code actually block because of this.

IIRC the situation in the test was roughly like this: The background networking thread caused a lot of RX resources to be blocked (queued) due to initial MLD, NS, ND, etc. But when the networking thread tried to allocate TX packages to handle the incoming packages it got starved as the test thread managed to block remaining TX resources in the meantime. Deadlock occurred because the test thread was waiting itself for some network event (assignment of an IP address or something - don't remember the details any more) which would then never happen.

Just pointing it out as I saw the network stack on real devices initially consume a lot of package resources exactly at the same time that userland processes gain control for the first time and may block TX resources in their own initialization routine while waiting for the network to become available.

Reserving a larger TX/RX thread will solve this but it's maybe not so obvious to users? Detecting the deadlock in the network thread (waiting for TX resources with timeout) and throwing a good error message may be a good enough solution then...

rlubos commented:
> > The thing I don't follow, is that net_if code makes no pkt/buffer allocation so it's not clear to me where could the net_if code actually block because of this.
> 
> IIRC the situation in the test was roughly like this: The background networking thread caused a lot of RX resources to be blocked (queued) due to initial MLD, NS, ND, etc. But when the networking thread tried to allocate TX packages to handle the incoming packages it got starved while holding on to the mutex as the test thread managed to block remaining TX resources in the meantime. Deadlock occurred because the test thread was waiting itself for some network event or needed some network resource protected by the mutex (some networking function call, RX or assignment of an IP address or something - don't remember the details any more) which would then never happen.
> 
> Just pointing it out as I saw the network stack on real devices initially consume a lot of package resources exactly at the same time that userland processes gain control for the first time and may block TX resources in their own initialization routine while waiting for the network to become available.
> 
> Reserving a larger TX/RX thread will solve this but it's maybe not so obvious to users? Detecting the deadlock in the network thread (waiting for TX resources with timeout) and throwing a good error message may be a good enough solution then...

On second thought, I think that this might've happen because of the loopback interface nature - it does not send the packet to the wire, but creates a **copy** of it (so it's an exception in terms of packet allocation) and sends it back to the stack. With `iface` buffer being held in `net_pkt_set_iface()` on the RX path (which please note is no longer the case after recent change), we may hit the RX packet shortage.

I think though that "deadlock" is a too strong description for the problem - yes, it may seems so as the whole boot sequence blocked for the 30 seconds, but this was due to DAD procedure failure. The actual impact of the mutex being held was packet allocation failure (there's a 100 ms timeout in the driver) so the packet did not reach back to the stack.

carlescufi reviewed (approved):
Reporter @fgrandel previously approved this PR before a rebase, so merging.
